functions/venue.js

```diff
@@ -476,23 +476,36 @@ exports.createVenue_v2 = functions.https.onCall(async (data, context) => {
 
 exports.upsertRoom = functions.https.onCall(async (data, context) => {
   checkAuth(context);
+
+  const result = {
+    provided: data,
+  };
+
   const { venueId, roomIndex, room } = data;
   await checkUserIsOwner(venueId, context.auth.token.user_id);
   const doc = await admin.firestore().collection("venues").doc(venueId).get();
 
   if (!doc || !doc.exists) {
     throw new HttpsError("not-found", `Venue ${venueId} not found`);
   }
-  const docData = doc.data();
-  let rooms = docData.rooms;
 
-  if (typeof roomIndex !== "number") {
-    rooms = [...rooms, room];
+  result.rooms = doc.data().rooms ?? [];
```

Comment:
I will take a look at it again. Few notes:

- I had one working version assigning the results differently It just added few more lines of code, didn't really change anything.
- The result of `doc.data()` isn't being mutated, it's just the removal of `docData` variable that makes it look like there's an issue.
- I'm unsure though about removing the `if` check, since I haven't gone that extensively through all the uses and the possible values therein. I'd leave it as is
- The idea of `result` being on top is to have visible cue the input params i.e. `data` are also attached to the returning result.

Reply:
Here is a refactor that has the original logic, is using spread, without mutation and without using dot operator to assign to results:

```javascript
  const docRooms = doc.data().rooms ?? [];

  const index = typeof roomIndex === "number" ? roomIndex : docRooms.length - 1;

  const rooms = [
    ...docRooms.slice(0, index),
    room,
    ...docRooms.slice(index + 1),
  ];

  const writeResult = await admin
    .firestore()
    .collection("venues")
    .doc(venueId)
    .update({ rooms });

  return {
    provided: data,
    roomIndex: index,
    rooms,
    writeResult,
  };
```

@denisdimitrov if you think this falls under the "would be great" definition, give it a go and I can make the change, otherwise I'd leave it as currently is.

Reply:
refactored in 1bb7aaa7c18844dca1014d68f039d8237c4bbf4d